### PR TITLE
Implement `MongoFuzzer`

### PR DESCRIPTION
**Pull request by tcoratger**
Time spent on this PR: 3h

Resolves: #748 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

- This PR introduces a fuzzer to generate MongoDB block header, block receipt and transaction documents randomly.
- The `MongoFuzzer` structure consumes an rnd and is capable of:
    - Add random transactions and receipts.
    - Add random block headers.
    - Finalize the structure and return an instance of the database.
    - Mock a complete database with a number of headers and transactions as input.

- The generation does not take into account any particular calculation between the fields (for example the elements of the signature v, r and s are false between them) but they are only generated for mock purposes to have random data available if necessary .

- Unit tests are added to check the basic correspondences between the elements of the transaction and the receipt.

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: src/eth_provider/database/mod.rs [hunk 1]
```diff
@@ -61,3 +61,23 @@ impl Database {
...
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Database {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
```
Comment (tcoratger) [inline src/eth_provider/database/mod.rs:77]: Should we do this that way or do we decide that we never use `deref` on wrappers? If so I will delete this and do it with another method.

File: src/eth_provider/database/types/header.rs [hunk 2]
```diff
@@ -7,3 +11,51 @@ pub struct StoredHeader {
     #[serde(deserialize_with = "crate::eth_provider::database::types::serde::deserialize_intermediate")]
     pub header: Header,
 }
+
+#[cfg(any(test, feature = "arbitrary"))]
+impl<'a> arbitrary::Arbitrary<'a> for StoredHeader {
```
Comment (greged93) [inline src/eth_provider/database/types/header.rs:16]: fire

File: src/eth_provider/database/mod.rs (hunk 1 above)
Comment (tcoratger) [reply src/eth_provider/database/mod.rs]: Just removed that here.

File: src/eth_provider/database/types/transaction.rs [hunk 3]
```diff
@@ -15,6 +19,57 @@ impl From<StoredTransaction> for Transaction {
...
+            access_list = Some(Default::default());
+            for item in &list.0 {
+                let address = item.address;
+                let storage_keys = &item.storage_keys;
+                access_list
+                    .as_mut()
+                    .unwrap()
+                    .push(reth_rpc_types::AccessListItem { address, storage_keys: storage_keys.to_vec() });
+            }
+        }
```
Comment (greged93) [inline src/eth_provider/database/types/transaction.rs:39]: I find this part a little strange, can't we use a `Vec<reth_rpc_types::AccessListItem>` and push the items in it? Or even just an `into_iter().map()` on `transaction.access_list()`?

File: src/models/transaction.rs [hunk 4]
```diff
@@ -70,6 +70,15 @@ pub fn rpc_transaction_to_primitive(
...
 
+// pub fn primitive_receipt_to_rpc(
+//     receipt: reth_primitives::Receipt,
+// ) -> Result<reth_rpc_types::TransactionReceipt, KakarotError> {
+//     Ok(reth_rpc_types::TransactionReceipt {
+
+//         transaction_hash: receipt
+//     })
+// }
+
```
Comment (greged93) [inline src/models/transaction.rs:81]: to remove?

File: src/test_utils/mongo/mod.rs [hunk 5]
```diff
@@ -1,13 +1,24 @@
...
-
+use crate::eth_provider::database::types::{
+    header::StoredHeader, receipt::StoredTransactionReceipt, transaction::StoredTransaction,
+};
 use crate::eth_provider::database::Database;
+#[cfg(any(test, feature = "arbitrary"))]
+use arbitrary::Arbitrary;
 use lazy_static::lazy_static;
+#[cfg(any(test, feature = "arbitrary"))]
+use mongodb::bson;
```
Comment (greged93) [inline src/test_utils/mongo/mod.rs:9]: Same here, I think you can group this with
```rust
#[cfg(any(test, feature = "arbitrary"))]
use {...}
```

File: src/test_utils/mongo/mod.rs [hunk 6]
```diff
@@ -1,13 +1,24 @@
...
 use mongodb::{
     bson::{doc, Document},
     options::{DatabaseOptions, ReadConcern, UpdateModifications, UpdateOptions, WriteConcern},
     Client, Collection,
 };
+#[cfg(any(test, feature = "arbitrary"))]
+use reth_primitives::U8;
 use reth_primitives::{constants::EMPTY_ROOT_HASH, Address, B256, U128, U256, U64};
+use serde::{Serialize, Serializer};
+#[cfg(any(test, feature = "arbitrary"))]
```
Comment (greged93) [inline src/test_utils/mongo/mod.rs:19]: same here

File: src/test_utils/mongo/mod.rs [hunk 7]
```diff
@@ -338,16 +349,248 @@ async fn update_many(doc: String, value: String, collection: Collection<Document
...
+    documents: HashMap<CollectionDB, Vec<StoredData>>,
+    /// Connection to the MongoDB database.
+    mongodb: Database,
+    /// Unstructured data
+    u: &'a mut arbitrary::Unstructured<'a>,
+}
+
+#[cfg(any(test, feature = "arbitrary"))]
+impl<'a> MongoFuzzer<'a> {
+    /// Creates a new instance of `MongoFuzzer` with the specified random number generator.
```
Comment (greged93) [inline src/test_utils/mongo/mod.rs:400]: I think comment needs an update

**Review by greged93 — COMMENTED** [review]
Couple of comments, also we need to make an issue just to replace the current tests qith the fuzzer :)

File: src/eth_provider/database/types/transaction.rs (hunk 3 above)
Comment (tcoratger) [reply src/eth_provider/database/types/transaction.rs]: Yes you are right, just simplified this part.

File: src/eth_provider/database/types/receipt.rs [hunk 8]
```diff
@@ -1,15 +1,79 @@
+#[cfg(any(test, feature = "arbitrary"))]
+use arbitrary::Arbitrary;
+#[cfg(any(test, feature = "arbitrary"))]
+use reth_primitives::{Address, Bloom, Receipt, B256, U128, U256, U64, U8};
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/database/types/receipt.rs]: Fixed

File: src/models/transaction.rs (hunk 4 above)
Comment (tcoratger) [reply src/models/transaction.rs]: Oups, yes fixed

File: src/test_utils/mongo/mod.rs (hunk 6 above)
Comment (tcoratger) [reply src/test_utils/mongo/mod.rs]: Fixed

File: src/test_utils/mongo/mod.rs (hunk 7 above)
Comment (tcoratger) [reply src/test_utils/mongo/mod.rs]: Fixed

File: src/test_utils/mongo/mod.rs (hunk 5 above)
Comment (tcoratger) [reply src/test_utils/mongo/mod.rs]: fixed

**Review by greged93 — APPROVED** [review]
lgtm!

**Review by greged93 — APPROVED** [review]
lgtm

**Pull request merged**